controllers/pod.go

```diff
@@ -713,3 +728,270 @@ func getPodNames(pods []*corev1.Pod) []string {
+						fileFound := false
+						if specStorage["files"] != nil {
+							for _, specFile := range specStorage["files"].([]interface{}) {
+								if specFile.(string) == statusFile.(string) {
+									fileFound = true
+									break
+								}
+							}
+						}
+
+						if !fileFound {
+							err := r.deleteFileStorage(pod, statusFile.(string))
+							if err != nil {
+								return err
+							}
+						}
+					}
+				}
+			}
+		}
+		if !namespaceFound {
+			r.Log.Info(
+				"namespace is deleted", "namespace", statusNamespace.(map[string]interface{})["name"],
+			)
+			statusStorage := statusNamespace.(map[string]interface{})["storage-engine"].(map[string]interface{})
+			if statusStorage["type"] == "device" {
+				if statusStorage["devices"] != nil {
+					for _, statusDevice := range statusStorage["devices"].([]interface{}) {
+						deviceName := getDeviceNameFromPath(rackStatus.Storage.Volumes, statusDevice.(string))
+						removedDevices = append(removedDevices, deviceName)
+					}
+					updateCluster = true
+				}
+				if statusStorage["files"] != nil {
+					for _, statusFile := range statusStorage["files"].([]interface{}) {
+						err := r.deleteFileStorage(pod, statusFile.(string))
+						if err != nil {
+							return err
+						}
+					}
+				}
+			}
+		}
+	}
+	if updateCluster {
+		var patches []jsonpatch.JsonPatchOperation
+
+		patch1 := jsonpatch.JsonPatchOperation{
+			Operation: "replace",
+			Path:      "/status/pods/" + pod.Name + "/dirtyVolumes",
+			Value:     removedDevices,
+		}
+
+		patches = append(patches, patch1)
+
+		jsonPatchJSON, err := json.Marshal(patches)
+		if err != nil {
+			return err
+		}
+		constantPatch := client.RawPatch(types.JSONPatchType, jsonPatchJSON)
+
+		// Since the pod status is updated from pod init container,
+		//set the field owner to "pod" for pod status updates.
+		if err = r.Client.Status().Patch(
+			context.TODO(), r.aeroCluster, constantPatch, client.FieldOwner("pod"),
+		); err != nil {
+			return fmt.Errorf("error updating status: %v", err)
+		}
+	}
+	return nil
+}
+
+func (r *SingleClusterReconciler) handleNSOrDeviceAddition(rackState RackState, podName string) (RestartType, error) {
+	var rackStatus asdbv1beta1.Rack
+	var err error
+
+	newAeroCluster := &asdbv1beta1.AerospikeCluster{}
+	err = r.Client.Get(
+		context.TODO(), types.NamespacedName{
+			Name: r.aeroCluster.Name, Namespace: r.aeroCluster.Namespace,
+		}, newAeroCluster,
+	)
+	if err != nil {
+		return QuickRestart, err
+	}
+
+	rackFound := false
+	for _, rackStatus = range newAeroCluster.Status.RackConfig.Racks {
+		if rackStatus.ID == rackState.Rack.ID {
+			rackFound = true
+			break
+		}
+	}
+
+	if !rackFound {
+		return QuickRestart, fmt.Errorf("could not find rack with ID: %d", rackState.Rack.ID)
+	}
+	r.Log.Info(
+		"Got rack status",
+	)
+
+	for _, specNamespace := range rackState.Rack.AerospikeConfig.Value["namespaces"].([]interface{}) {
```

Comment:
Please check if it is possible to move this complete check outside so that computation should run once and reused for all pods